examples/linalg/src/linalg.rs

```diff
@@ -20,52 +20,252 @@
 
 use hacspec_lib::*;
 
-pub type DimType = usize;
-pub type Scalar = i128;
-pub type Dims = (DimType, DimType);
-pub type Matrix = (Dims, Seq<Scalar>);
-pub type MatRes = Result<Matrix, ()>;
-
-// Generate new matrix using rows, cols and a seq.
-pub fn new(rows: DimType, cols: DimType, seq: Seq<Scalar>) -> MatRes {
+type DimType = usize;
+type Dims = (DimType, DimType);
+type MatRes<T> = Result<Matrix<T>, u8>;
+type ScalRes<T> = Result<T, u8>;
+
+pub type Matrix<T> = (Dims, Seq<T>);
+
+// === Errors === //
+
+const DIMENSION_SEQUENCE_LENGTH_MISMATCH: u8 = 10u8;
+const INDEX_OUT_OF_BOUNDS: u8 = 11u8;
+const SLICE_OUT_OF_BOUNDS: u8 = 12u8;
+const DIMENSION_MISMATCH: u8 = 13u8;
+
+// === External Functions === //
+
+/// Generate new matrix using rows, cols and a seq. Returns an error
+/// if the product of the given dimensions is larger than the length
+/// of the given Seq.
+pub fn new<T>(rows: DimType, cols: DimType, seq: Seq<T>) -> MatRes<T>
+where
+    T: hacspec_lib::Integer,
+{
     if seq.len() > 0 && rows * cols == seq.len() {
         MatRes::Ok(((rows, cols), seq))
     } else {
-        MatRes::Err(())
+        MatRes::Err(DIMENSION_SEQUENCE_LENGTH_MISMATCH)
     }
 }
 
-// Generate a n*m matrix filled with a given scalar
-pub fn repeat(n: DimType, m: DimType, scalar: Scalar) -> MatRes {
-    let mut ret = Seq::<Scalar>::new(n * m);
+/// Generate a n*m matrix filled with a given scalar.
+pub fn repeat<T>(n: DimType, m: DimType, scalar: T) -> Matrix<T>
+where
+    T: hacspec_lib::Integer,
```

Comment:
I see that we need this here. But it would be nice if we could get away without `where` clauses.
The question is what is the easiest and tightest bound we could put on `T` in general here.